crates/red_knot/src/semantic/types.rs

```diff
@@ -234,7 +236,14 @@ impl TypeStore {
     }
 
     fn add_union(&self, file_id: FileId, elems: &[Type]) -> UnionTypeId {
-        self.add_or_get_module(file_id).add_union(elems)
+        let flattened: Vec<Type> = elems
+            .iter()
+            .flat_map(|ty| match ty {
+                Type::Union(union_id) => Left(union_id.elements(self)),
+                _ => Right(std::iter::once(*ty)),
+            })
+            .collect();
```

Comment:
It took me a little while to understand why `either` was being used here. I think here I'd find a more imperative approach more readable (and then we don't need the new `either` dependency for the crate):

```suggestion
        let mut flattened = Vec::with_capacity(elems.len());
        for elem in elems {
            match elem {
                Type::Union(union_id) => flattened.extend(union_id.elements(self)),
                _ => flattened.push(*elem),
            }
        }
```

Reply:
Ugh, but it took me a while to figure out how to do this successfully with `flat_map`! Ok, ok, I'll switch to imperative style...